### ISSUE TITLE
Compatibility issue with pytorch v1.7.0 and pose SGD

### ISSUE THREAD

**Issue by Icecream-blue-sky** [zhonge/cryodrgn#29]
When I run "cryodrgn train_nn data/toy_projections.mrcs --poses data/toy_rot_zerotrans.pkl -o output/toy_recon --do-pose-sgd
", I get the "ValueError: optimizer got an empty parameter list". The whole log is below:
```shell
ZOU@LAPTOP-GMM4MN25 MINGW64 /d/Study/The Road to AI/AIRoad/cryodrgn/testing (master)
$ cryodrgn train_nn data/toy_projections.mrcs --poses data/toy_rot_zerotrans.pkl -o output/toy_recon --do-pose-sgd
2020-12-14 09:27:02     Namespace(amp=False, batch_size=8, checkpoint=1, ctf=None, datadir=None, dim=256, do_pose_sgd=True, domain='fourier', emb_typ
e='quat', func=<function main at 0x000001E7D44836A8>, ind=None, invert_data=True, l_extent=0.5, layers=3, lazy=False, load=None, log_interval=1000, l
r=0.0001, multigpu=False, norm=None, num_epochs=20, outdir='D:\\Study\\The Road to AI\\AIRoad\\cryodrgn\\testing\\output\\toy_recon', particles='D:\\
Study\\The Road to AI\\AIRoad\\cryodrgn\\testing\\data\\toy_projections.mrcs', pe_dim=None, pe_type='geom_lowf', pose_lr=0.0001, poses='D:\\Study\\Th
e Road to AI\\AIRoad\\cryodrgn\\testing\\data\\toy_rot_zerotrans.pkl', pretrain=5, relion31=False, seed=95562, verbose=False, wd=0, window=True)
2020-12-14 09:27:02     Use cuda True
2020-12-14 09:27:02     Loaded 1000 30x30 images
2020-12-14 09:27:02     Normalized HT by 0 +/- 25.786991119384766
2020-12-14 09:27:06     FTPositionalDecoder(
  (decoder): ResidLinearMLP(
    (main): Sequential(
      (0): Linear(in_features=90, out_features=256, bias=True)
      (1): ReLU()
      (2): ResidLinear(
        (linear): Linear(in_features=256, out_features=256, bias=True)
      )
      (3): ReLU()
      (4): ResidLinear(
        (linear): Linear(in_features=256, out_features=256, bias=True)
      )
      (5): ReLU()
      (6): ResidLinear(
        (linear): Linear(in_features=256, out_features=256, bias=True)
      )
      (7): ReLU()
      (8): Linear(in_features=256, out_features=2, bias=True)
    )
  )
)
2020-12-14 09:27:06     221186 parameters in model
Traceback (most recent call last):
  File "D:\Anaconda\envs\cryodrgn\Scripts\cryodrgn-script.py", line 33, in <module>
    sys.exit(load_entry_point('cryodrgn==0.3.0b0', 'console_scripts', 'cryodrgn')())
  File "D:\Anaconda\envs\cryodrgn\lib\site-packages\cryodrgn-0.3.0b0-py3.7.egg\cryodrgn\__main__.py", line 50, in main
    args.func(args)
  File "D:\Anaconda\envs\cryodrgn\lib\site-packages\cryodrgn-0.3.0b0-py3.7.egg\cryodrgn\commands\train_nn.py", line 169, in main
    pose_optimizer = torch.optim.SparseAdam(posetracker.parameters(), lr=args.pose_lr)
  File "D:\Anaconda\envs\cryodrgn\lib\site-packages\torch\optim\sparse_adam.py", line 49, in __init__
    super(SparseAdam, self).__init__(params, defaults)
  File "D:\Anaconda\envs\cryodrgn\lib\site-packages\torch\optim\optimizer.py", line 47, in __init__
    raise ValueError("optimizer got an empty parameter list")
ValueError: optimizer got an empty parameter list
```
I find it caused by the code in /cryodrgn/commands/train_nn.py:
```python
 # load poses
    if args.do_pose_sgd:
        posetracker = PoseTracker.load(args.poses, Nimg, D, args.emb_type, ind)
        pose_optimizer = torch.optim.SparseAdam(posetracker.parameters(), lr=args.pose_lr)
```
If I try to run "train_vae" with the same setting, I get the same error caused by the same code in  /cryodrgn/commands/train_vae.py:
```python
  # load poses
    if args.do_pose_sgd: assert args.domain == 'hartley', "Need to use --domain hartley if doing pose SGD"
    do_pose_sgd = args.do_pose_sgd
    posetracker = PoseTracker.load(args.poses, Nimg, D, 's2s2' if do_pose_sgd else None, ind)
    pose_optimizer = torch.optim.SparseAdam(posetracker.parameters(), lr=args.pose_lr) if do_pose_sgd else None
```
So I think it may be an undiscovered bug...
My environment setting is below:
Win10, PyCharm, cryodrgn 0.3.0 b

**Comment by XiangmengCai** [comment]
Hi @Icecream-blue-sky,

I had the same issue before using train_vae.py. A simple fix would be to use
```
  # load poses
    if args.do_pose_sgd: assert args.domain == 'hartley', "Need to use --domain hartley if doing pose SGD"
    do_pose_sgd = args.do_pose_sgd
    posetracker = PoseTracker.load(args.poses, Nimg, D, 's2s2' if do_pose_sgd else None, ind)
    pose_optimizer = torch.optim.SparseAdam(list(posetracker.parameters()), lr=args.pose_lr) if do_pose_sgd else None
```
It solved the problem for me and probably would be the same for train_nn.py. Hope it helps!

**Comment by Icecream-blue-sky** [comment]
> Hi @Icecream-blue-sky,
> 
> I had the same issue before using train_vae.py. A simple fix would be:
> 
> ```
>   # load poses
>     if args.do_pose_sgd: assert args.domain == 'hartley', "Need to use --domain hartley if doing pose SGD"
>     do_pose_sgd = args.do_pose_sgd
>     posetracker = PoseTracker.load(args.poses, Nimg, D, 's2s2' if do_pose_sgd else None, ind)
>     pose_optimizer = torch.optim.SparseAdam(list(posetracker.parameters()), lr=args.pose_lr) if do_pose_sgd else None
> ```
> 
> It solved the problem for me and probably would be the same for train_nn.py. Hope it helps!

Thanks a lot. It also works for me! But I think this problem is still really weird. In  file "D:\Anaconda\envs\cryodrgn\lib\site-packages\torch\optim\optimizer.py":
```python
    def __init__(self, params, defaults):
        torch._C._log_api_usage_once("python.optimizer")
        self.defaults = defaults
        if isinstance(params, torch.Tensor):
            raise TypeError("params argument given to the optimizer should be "
                            "an iterable of Tensors or dicts, but got " +
                            torch.typename(params))

        self.state = defaultdict(dict)
        self.param_groups = []
        param_groups = list(params)
        if len(param_groups) == 0:
            raise ValueError("optimizer got an empty parameter list")
        if not isinstance(param_groups[0], dict):
            param_groups = [{'params': param_groups}]

        for param_group in param_groups:
            self.add_param_group(param_group)
```
posetracker.parameters() is assigned to var params and then become list(params) through
```python
param_groups = list(params)
```
which means that posetracker.parameters() will become list(posetracker.parameters()) automatically after torch.optim.SparseAdam(posetracker.parameters()...). So why I still need to add list(posetracker.parameters()) ?

**Comment by zhonge** [comment]
Thanks @XiangmengCai for pointing this out. The command runs fine for me -- what version of pytorch are you using?

**Comment by XiangmengCai** [comment]
Thanks for replying @zhonge. I'm using Pytorch 1.7.0.

**Comment by zhonge** [comment]
Fixed in commit cd365ad85338044e21ac7a1b22350cd39974870e. Thanks all!
